src/analyses/base.ml

```diff
@@ -2026,6 +2026,80 @@ struct
         M.warn ~category:(Behavior (Undefined InvalidMemoryDeallocation)) ~tags:[CWE 761] "Free of memory not at start of buffer in function %s for pointer %a" special_fn.vname d_exp ptr
     | _ -> M.warn ~category:MessageCategory.Analyzer "Pointer %a in function %s doesn't evaluate to a valid address." d_exp ptr special_fn.vname
 
+  (* Get the size of the smallest memory that dest points-to *)
+  let get_min_size_of_dest ctx dest =
+    let intdom_of_int x =
+      ID.of_int (Cilfacade.ptrdiff_ikind ()) (Z.of_int x)
+    in
+    let points_to_heap_only =
+      match ctx.ask (Queries.MayPointTo dest) with
+      | a when not (AD.is_top a) ->
+        AD.for_all (function
+            | Addr (v, _) -> ctx.ask (Queries.IsHeapVar v)
+            | _ -> false
+          ) a
+      | _ -> false
+    in
+    if points_to_heap_only then
+      (* Ask for BlobSize from the base address (the second field set to true) in order to avoid BlobSize giving us bot *)
+      ctx.ask (Queries.BlobSize {exp = dest; base_address = true})
+    else
+      match ctx.ask (Queries.MayPointTo dest) with
+      | a when not (Queries.AD.is_top a) ->
+        let pts_list = Queries.AD.elements a in
+        let pts_elems_to_sizes (addr: Queries.AD.elt) =
+          begin match addr with
+            | Addr (v, _) ->
+              begin match v.vtype with
+                | TArray (item_typ, _, _) ->
+                  let item_typ_size_in_bytes = (bitsSizeOf item_typ) / 8 in
+                  let item_typ_size_in_bytes = intdom_of_int item_typ_size_in_bytes in
+                  begin match ctx.ask (Queries.EvalLength dest) with
+                    | `Lifted arr_len -> `Lifted (ID.mul item_typ_size_in_bytes arr_len)
+                    | `Bot -> `Bot
+                    | `Top -> `Top
+                  end
+                | _ ->
+                  let type_size_in_bytes = (bitsSizeOf v.vtype) / 8 in
+                  `Lifted (intdom_of_int type_size_in_bytes)
+              end
+            | _ -> `Top
+          end
+        in
+        (* Map each points-to-set element to its size *)
+        let pts_sizes = List.map pts_elems_to_sizes pts_list in
+        (* Take the smallest of all sizes that ptr's contents may have *)
+        begin match pts_sizes with
+          | [] -> `Bot
+          | [x] -> x
+          | x::xs -> List.fold_left (fun acc elem ->
+              if ValueDomainQueries.ID.compare acc elem >= 0 then elem else acc
+            ) x xs
```

Comment:
I don't think this is correct here, comparison of ID.t values does not need to coincide with any sort of order. The better thing here would be `ID.join`.

Reply:
Changed the `fold_left` to use `join` accumulating the sizes together in https://github.com/goblint/analyzer/pull/1197/commits/655362e70a7d0c2a42c4dba910e5c67de1aee1c9.